server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1284,37 +1290,47 @@ void run(ActionListener<Void> listener) {
 
             private void determineShardCount(ActionListener<Void> listener) {
                 try (var listeners = new RefCountingListener(listener)) {
-                    for (final var indexMetaGeneration : snapshotIds.stream()
+                    for (final var blobId : snapshotIds.stream()
                         .filter(snapshotsWithIndex::contains)
                         .map(id -> originalRepositoryData.indexMetaDataGenerations().indexMetaBlobId(id, indexId))
                         .collect(Collectors.toSet())) {
-                        // NB since 7.9.0 we deduplicate index metadata blobs, and one of the components of the deduplication key is the
-                        // index UUID; the shard count is going to be the same for all metadata with the same index UUID, so it is
-                        // unnecessary to read multiple metadata blobs corresponding to the same index UUID.
-                        // TODO Skip this unnecessary work? Maybe track the shard count in RepositoryData?
-                        snapshotExecutor.execute(ActionRunnable.run(listeners.acquire(), () -> getOneShardCount(indexMetaGeneration)));
+                        snapshotExecutor.execute(ActionRunnable.run(listeners.acquire(), () -> {
+                            // NB since 7.9.0 we deduplicate index metadata blobs, and one of the components of the deduplication key is the
+                            // index UUID; the shard count is going to be the same for all metadata with the same index UUID, so it is
+                            // unnecessary to read multiple metadata blobs corresponding to the same index UUID.
+                            String indexUUID = originalRepositoryData.indexMetaDataGenerations().convertBlobIdToIndexUUID(blobId);
+                            if (indexUUIDToShardCountMap.containsKey(indexUUID) == false) {
```

Comment:
This still risks duplicates since we call `containsKey` separately from `put`. Once we're just tracking the set of the index UUIDs we can just call `add` here which returns a flag indicating whether the item was newly added.